File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque


def solution(maps):
    moves = ((-1, 0), (1, 0), (0, -1), (0, 1))
    que = deque()
    que.append((0, 0, 1))
    visited = set()
    visited.add((0, 0))

    while que:
        r, c, count = que.popleft()

        if r == len(maps) - 1 and c == len(maps[0]) - 1:
            return count

        for dr, dc in moves:
            nr, nc = dr + r, dc + c

            if (0 <= nr < len(maps)) and (0 <= nc < len(maps[0])) and maps[nr][nc] == 1 and (nr, nc) not in visited:
                que.append((nr, nc, count + 1))
                visited.add((nr, nc))
    
    return -1

File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    function bfs() {
        const que = [[0, 0, 1]]
        visited[0][0] = 1
        
        while (que.length) {
            const [r, c, count] = que.shift()
            
            if (r === n - 1 && c === m - 1) {
                return count
            }
            
            for (const [dr, dc] of moves) {
                const nr = r + dr,
                      nc = c + dc
                if ((0 <= nr && nr < n) && 
                    (0 <= nc && nc < m) && 
                    visited[nr][nc] === 0 && 
                    maps[nr][nc] === 1) 
                {
                    que.push([nr, nc, count + 1])
                    visited[nr][nc] = 1
                }
            }
        }
        
        return -1
    }
    
    const n = maps.length,
          m = maps[0].length
    const moves = [[-1, 0], [1, 0], [0, -1], [0, 1]]
    const visited = Array(n).fill().map(()=> Array(m).fill(0))
    return bfs();
}